### PR TITLE
Add validators

### PR DESCRIPTION
**Pull request by keito-jp**
Add following validators

- minimum
- max length
- minimum length
- pattern

### CODE REVIEW

File: lib/index.ts [hunk 1]
```diff
@@ -2,3 +2,7 @@ export * from "./errors";
 export * from "./interfaces";
 export * from "./validators/enum";
```
Comment (minodisk) [inline lib/index.ts]: Should export `./validators/format`

**Comment by minodisk** [comment]
Should implement other validators for array:

- `minItems`
- `maxItems`

**Comment by minodisk** [comment]
Rename files:

- maxitems.ts -> max_items.ts
- minitems.ts -> min_items.ts
- maxlength.ts -> max_length.ts
- minlength.ts -> min_length.ts
- maxitems_test.ts -> max_items_test.ts
- minitems_test.ts -> min_items_test.ts

**Comment by keito-jp** [comment]
@minodisk done!

File: lib/validators/format.ts [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      default:
+        throw new InvalidFormatError(`the format is not found`);
+    }
+  }
+
+  public validate(input: string): IValidationError<string, IFormatValidatorDefinition> {
+    const {format} = this.definition;
+    switch (format) {
+      case "date-time": {
+        const regExp: RegExp = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
```
Comment (minodisk) [inline lib/validators/format.ts]: Should construct `RegExp` instance as static const.

File: lib/validators/max_items.ts [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+export interface IMaxItemsValidatorDefinition {
+  maxItems: number;
+}
+
+export class MaxItemsValidator implements IValidator<string, IMaxItemsValidatorDefinition> {
+  constructor(public definition: IMaxItemsValidatorDefinition) {
+    const {maxItems} = this.definition;
+    if (maxItems < 0) {
+     throw new NotLengthError(`the value of maxItems should be greater than, or equal to, 0`);
```
Comment (minodisk) [inline lib/validators/max_items.ts]: Rename to `NoLengthError`

File: lib/index.ts (hunk 1 above)
Comment (minodisk) [inline lib/index.ts]: Is this file linted?

Comment (keito-jp) [inline lib/index.ts]: yep

File: lib/validators/max_length.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+  public validate(input: string): IValidationError<string, IMaxLengthValidatorDefinition> {
+    if (input.length <= this.definition.maxLength) {
+      return;
+    }
+    return {
+      input,
+      definition: this.definition,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/max_length.ts]: LGTM

File: lib/validators/min_items.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+  public validate(input: any[]): IValidationError<string, IMinItemsValidatorDefinition> {
+    if (input.length >= this.definition.minItems) {
+      return;
+    }
+    return {
+      input,
+      definition: this.definition,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/min_items.ts]: LGTM

File: lib/validators/min_length.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  public validate(input: string): IValidationError<string, IMinLengthValidatorDefinition> {
+    if (input.length >= this.definition.minLength) {
+      return;
+    }
+    return {
+      input,
+      definition: this.definition,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/min_length.ts]: LGTM

File: lib/validators/max_items.ts [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+import {
+  IValidationError,
+  IValidator,
+} from "../interfaces";
+
+export interface IMaxItemsValidatorDefinition {
+  maxItems: number;
+}
+
+export class MaxItemsValidator implements IValidator<string, IMaxItemsValidatorDefinition> {
```
Comment (minodisk) [inline lib/validators/max_items.ts]: The `maxItems` shouldn't check the length of `string` but `array`.

File: lib/validators/min_items.ts [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import {
+  IValidationError,
+  IValidator,
+} from "../interfaces";
+
+export interface IMinItemsValidatorDefinition {
+  minItems: number;
+}
+
+export class MinItemsValidator implements IValidator<string, IMinItemsValidatorDefinition> {
```
Comment (minodisk) [inline lib/validators/min_items.ts]: The `MinItemsValidator` shouldn't check the length of `string` but `array`.

File: lib/validators/minimum.ts [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+  IValidator,
+} from "./interfaces";
+
+export interface IMinimumValidatorDefinition {
+  minimum: number;
+  exclusive: boolean;
+}
+
+export class MinimumValidator implements IValidator<number, IMinimumValidatorDefinition> {
+  constructor(public definition: IMinimumValidatorDefinition) { }
```
Comment (minodisk) [inline lib/validators/minimum.ts]: Delete white space.

File: lib/validators/pattern.ts [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+
+  constructor(public definition: IPatternValidatorDefinition) {
+    const {pattern} = this.definition;
+    if (pattern === "") {
+      throw new EmptyError(`the pattern should not be empty`);
+    }
+    try {
+      this.regExp = new RegExp(pattern);
+    } catch (e) {
+      throw new InvalidPatternError(`invalid pattern: ` + e);
```
Comment (minodisk) [inline lib/validators/pattern.ts]: Shorten `invalid pattern: ${e.toString()}`

File: lib/validators/pattern.ts [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    }
+  }
+
+  public validate(input: string): IValidationError<string, IPatternValidatorDefinition> {
+    const ok = this.regExp.test(input);
+    if (ok) {
```
Comment (minodisk) [inline lib/validators/pattern.ts]: Shorten to `if (this.regExp.test(input))`

File: lib/validators/required.ts [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    const {required} = this.definition;
+
+    for (let i = 0; i < required.length; i++) {
+      const value = required[i];
+      const hasKey: boolean = false;
+      Object.keys(input).forEach((key) => {
+        if (key === value) {
+          hasKey = true;
+        }
+      });
```
Comment (minodisk) [inline lib/validators/required.ts]: Wasted loop.
Just do this to check that an `Object` has the key:

```typescript
const key = required[i];
if (input[key] == null) {
  return someError;
}
```

**Comment by keito-jp** [comment]
@minodisk I fixed code where you point out.

File: lib/validators/format.ts [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+          return;
+        }
+        break;
+    }
+    return {
+      definition: this.definition,
+      input,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/format.ts]: LGTM

File: lib/validators/max_items.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+  public validate(input: any[]): IValidationError<any[], IMaxItemsValidatorDefinition> {
+    if (input.length <= this.definition.maxItems) {
+      return;
+    }
+    return {
+      input,
+      definition: this.definition,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/max_items.ts]: LGTM

File: lib/validators/min_items.ts [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+  public validate(input: any[]): IValidationError<any[], IMinItemsValidatorDefinition> {
+    if (input.length >= this.definition.minItems) {
+      return;
+    }
+    return {
+      input,
+      definition: this.definition,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/min_items.ts]: LGTM

File: lib/validators/minimum.ts [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+
+    if (input > this.definition.minimum) {
+      return;
+    }
+    return {
+      input,
+      definition: this.definition,
+    };
+  }
+}
```
Comment (minodisk) [inline lib/validators/minimum.ts]: LGTM

File: lib/validators/pattern.ts [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+
+  constructor(public definition: IPatternValidatorDefinition) {
+    const {pattern} = this.definition;
+    if (pattern === "") {
+      throw new EmptyError(`the pattern should not be empty`);
+    }
+    try {
+      this.regExp = new RegExp(pattern);
+    } catch (e) {
+      throw new InvalidPatternError(`invalid pattern: $(e.toString)`);
```
Comment (minodisk) [inline lib/validators/pattern.ts]: Typo: `${e.toString}`

File: lib/validators/required.ts [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+        return {
+          definition: this.definition,
+          input,
+        };
+      }
+    }
+
+    return;
+  }
+}
```
Comment (minodisk) [inline lib/validators/required.ts]: LGTM

**Comment by keito-jp** [comment]
@minodisk Fixed!

File: lib/validators/required.ts [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+export class RequiredValidator implements IValidator<Object, IRequiredValidatorDefinition> {
+  constructor(public definition: IRequiredValidatorDefinition) {
+    const {required} = this.definition;
+    const len = required.length;
+    if (len === 0) {
+      throw new EmptyError(`the required value should have at least one element`);
+    }
+
+    for (let i = 0; i < len - 1; i++) {
+      const e = required[i];
```
Comment (minodisk) [inline lib/validators/required.ts]: Don't use `e` but `key`.

File: lib/validators/max_items.ts [hunk 20]
```diff
@@ -0,0 +1,31 @@
+import {
+  NoLengthError,
+} from "../errors";
+
```
Comment (minodisk) [inline lib/validators/max_items.ts]: Delete blank line.

File: lib/validators/min_items.ts [hunk 21]
```diff
@@ -0,0 +1,31 @@
+import {
+  NoLengthError,
+} from "../errors";
+
```
Comment (minodisk) [inline lib/validators/min_items.ts]: Delete blank line.

File: lib/validators/format.ts [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+} from "../interfaces";
+
+export interface IFormatValidatorDefinition {
+  format: string;
+}
+
+export class FormatValidator implements IValidator<string, IFormatValidatorDefinition> {
+  private static rDateTime: RegExp = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}Z$/;
+  private static rEmail: RegExp = /^.+@.+\..+$/;
+  private static rUri: RegExp = /^https?:\/\/.+$/;
```
Comment (minodisk) [inline lib/validators/format.ts]: Should allow any protocol.

File: lib/index.ts [hunk 23]
```diff
@@ -1,4 +1,12 @@
...
+export * from "./validators/format";
+export * from "./validators/max_length";
+export * from "./validators/max_items";
 export * from "./validators/maximum";
+export * from "./validators/min_length";
+export * from "./validators/min_items";
+export * from "./validators/minimum";
+export * from "./validators/pattern";
+export * from "./validators/required";
```
Comment (minodisk) [inline lib/index.ts]: Add new line to the end of file.

**Pull request merged**